apollo-federation/src/sources/connect/id.rs

```diff
@@ -0,0 +1,118 @@
+pub(crate) enum ConnectedElement<'schema> {
+    Field {
+        parent_type: &'schema ExtendedType,
+        field_def: &'schema Component<FieldDefinition>,
+    },
+    Type {
+        #[allow(unused)]
+        type_def: &'schema ExtendedType,
+    },
+}
+
+impl ConnectorPosition {
+    pub(crate) fn element<'s>(
+        &self,
+        schema: &'s Schema,
+    ) -> Result<ConnectedElement<'s>, FederationError> {
+        match self {
+            Self::Field(pos) => Ok(ConnectedElement::Field {
+                parent_type: schema.types.get(pos.field.parent().type_name()).ok_or(
+                    FederationError::internal("Parent type for connector not found"),
+                )?,
+                field_def: pos.field.get(schema).map_err(|_| {
+                    FederationError::internal("Field definition for connector not found")
+                })?,
+            }),
+            Self::Type(pos) => Ok(ConnectedElement::Type {
+                type_def: schema
+                    .types
+                    .get(&pos.type_name)
+                    .ok_or(FederationError::internal("Type for connector not found"))?,
+            }),
+        }
+    }
+
+    pub(crate) fn parent_type_name(&self) -> Option<Name> {
+        match self {
+            ConnectorPosition::Field(pos) => Some(pos.field.type_name().clone()),
+            ConnectorPosition::Type(_) => None,
+        }
+    }
+
+    pub(crate) fn root_type_name(&self, schema: &Schema) -> Option<NamedType> {
+        match self {
+            ConnectorPosition::Field(_) => self
+                .field_definition(schema)
+                .map(|field| field.ty.inner_named_type().clone()),
+            ConnectorPosition::Type(pos) => Some(pos.type_name.clone()),
+        }
+    }
+
+    pub(crate) fn field_definition<'s>(
+        &self,
+        schema: &'s Schema,
+    ) -> Option<&'s Component<FieldDefinition>> {
+        match self {
+            ConnectorPosition::Field(pos) => pos.field.get(schema).ok(),
+            ConnectorPosition::Type(_) => None,
+        }
+    }
+
+    pub(crate) fn coordinate(&self) -> String {
+        match self {
+            ConnectorPosition::Field(pos) => format!(
+                "{}.{}@{}[{}]",
+                pos.field.type_name(),
+                pos.field.field_name(),
+                pos.directive_name,
+                pos.directive_index,
+            ),
+            ConnectorPosition::Type(pos) => format!(
+                "{}@{}[{}]",
+                pos.type_name, pos.directive_name, pos.directive_index,
+            ),
+        }
+    }
+
+    pub(crate) fn synthetic_name(&self) -> String {
+        match self {
+            ConnectorPosition::Field(pos) => format!(
+                "{}_{}_{}",
+                pos.field.type_name(),
+                pos.field.field_name(),
+                pos.directive_index,
+            ),
+            ConnectorPosition::Type(pos) => format!("{}_{}", pos.type_name, pos.directive_index,),
```

Comment:
```suggestion
            ConnectorPosition::Type(pos) => format!("{}_{}", pos.type_name, pos.directive_index),
```
When are they going to teach `rustfmt` about `format!`? 😆